### PR TITLE
Various fixes to improve automation-tools debugging, and create more seamless execution

### PR DESCRIPTION
**Pull request by ross-spencer**
This PR is in support of the work being done to debug the automation-tools workflow at the University of Melbourne where a number of issues have been observed, but cannot all be pinned down accurately without automation-tools performing better. 

* We now supply the config_file argument consistently to all run_scripts functions.
* Exception handling has been made more specific to ensure tha other errors are not masked during execution.
* A post-execution hook has been implemented to clean-up the script once it has successfully exited, e.g. to ensure that the process lock pid.lck file is removed without having to explicitly make that call more than once in the code.
* The post-execution hook will also look after the session closure for the database.
* Logging has been improved where possible to either shorten output not being directed to a DEBUG log, or make messages clearer or more specific when sent elsewhere.
* Transfers with duplicate names which are renamed in Archivematica can now be handled by the Automation Tools. 
* A unit test has been written to test this function against the API. 
* Database functions have been moved to the `models.py` module to separate the concerns of the code and to provide greater visibility of the work being done to talk to the database. This change should open up the opportunity to make the data stored in the database more expressive. 
* An SQLAlchemy data type renamed in the SQLAlchemy library has been renamed according to the tox warnings.

Connected to archivematica/issues#362
Connected to archivematica/issues#365
Connected to archivematica/issues#366
Connected to archivematica/issues#372
Connected to archivematica/issues#381
Connected to archivematica/issues#395
Connected to https://github.com/artefactual/automation-tools/issues/5

There are three parts to this which affect https://github.com/artefactual/automation-tools/issues/5, 1. the behavior couldn't be recreated; 2. there should be less risk with greater number of atomic commits that there will be any negative impact of a database connection being left open too long; 3. The session is now closed by the automation tools where previously this wasn't an explicit call made in the code.

### CODE REVIEW

File: transfers/transfer.py [hunk 1]
```diff
@@ -143,20 +165,28 @@ def get_accession_id(dirname):
...
+    except ValueError as err:
+        LOGGER.warning(
+            "ValueError: Unable to parse output from %s. Output: %s. %s",
+            script_path, output, err)
+        return None
+    except SyntaxError as err:
+        LOGGER.warning(
+            "SyntaxError: Unable to parse output from %s. Output: %s. %s",
+            script_path, output, err)
         return None
```
Comment (sevein) [inline transfers/transfer.py]: These are so similar that you could just `except (ValueError, SyntaxError) as err: ...`.

File: transfers/models.py [hunk 2]
```diff
@@ -34,3 +39,87 @@ def init(databasefile):
...
+    """
+    return session.query(Unit).\
+        filter_by(unit_type=unit_type, uuid=unit_uuid).one()
+
+
+def _update_unit(session, uuid, path, unit_type, status, current):
+    """Internal function to handle the updating of a unit in the database as
+    a single atomic transaction.
+    """
+    unit_ = Unit(
```
[the remark this answers is not in the document]
Comment (ross-spencer) [reply transfers/models.py]: I just liked it vs. Unit the object. Easy enough to change to `unit`.

File: transfers/transfer.py [hunk 3]
```diff
@@ -574,18 +571,16 @@ def main(am_user, am_api_key, ss_user, ss_api_key, ts_uuid, ts_path, depth,
...
     # If failed, rejected, completed etc, start new transfer
     if current_unit:
-        current_unit.current = False
+        models.update_unit_current(session, current_unit, False)
     new_transfer = start_transfer(
         ss_url, ss_user, ss_api_key, ts_uuid, ts_path,
         depth, am_url, am_user, am_api_key,
         transfer_type, see_files, session, config_file)
-
-    session.commit()
```
[the remark this answers is not in the document]
Comment (ross-spencer) [reply transfers/transfer.py]: `start_transfer` now takes responsibility for calling calling the following:

```
models.get_completed_transfers(...)
models.add_new_transfer(...)
models.failed_to_approve(...)
```

Each of which will update the unit and call commit if the code reaches them. There may be an optimization to be found here, but it does mean we shouldn't be concerned now about the commit absolutely having to happen at the end of the main thread of execution. 

I don't expect the session to be closed until after this is called, so I think we're good.

File: transfers/transfer.py [hunk 4]
```diff
@@ -311,13 +311,53 @@ def get_next_transfer(ss_url, ss_user, ss_api_key, ts_location_uuid,
     return None
 
 
+def call_start_transfer_endpoint(
+        am_url, am_user, am_api_key, target, transfer_type, accession,
+        ts_location_uuid, session):
+    """Make the call to the start_transfer endpoint and return the unapproved
+    directory name.
+    """
```
[the remark this answers is not in the document]
Comment (ross-spencer) [reply transfers/transfer.py]: I considered this when refactoring this and feel that making this a responsibility of AM Client is beyond the scope of what I can do with this ticket here. Definitely a change we should make in the future. 

I'll take a look at the second point.

File: tests/test_transfers.py [hunk 5]
```diff
@@ -181,18 +181,17 @@ def test_get_next_transfer_all_complete(self):
...
         # Set bad TS Location UUID
         ts_location_uuid = 'badd8d39-9cee-495e-b7ee-5e6292549bad'
         # Test
         path = transfer.get_next_transfer(SS_URL, SS_USER, SS_KEY,
                                           ts_location_uuid, PATH_PREFIX,
                                           DEPTH, COMPLETED, FILES)
-        # Verify
-        self.assertEqual(path,
-                         errors.error_lookup(errors.ERR_INVALID_RESPONSE))
+        # Verify that None is returned from the function.
```
[the remark this answers is not in the document]
Comment (ross-spencer) [reply tests/test_transfers.py]: Nice pic! :smile: 

Changed, and can consider how to make my comments more useful in future. The more `transfer.py` is refactored, the clearer the execution path for other readers will become, so comments can be used more sparingly with more impact.

**Comment by ross-spencer** [comment]
Hi @sevein I've made some changes which should be ready to review. Thanks for the first go at that.

I've had to be as pragmatic to the client as possible for the amount of time I reasonably have to improve this. If there are any non-major issues, but that might require another hour or so dev time, then I am hoping that we can open another GitHub issue and come back to it later.

As well as improving the unit-tests I have been testing an AT workflow locally to try and captureas many different error conditions as possible. You probably saw part of that in the last commits. Hopefully things have been pushed forward a bit.

**Pull request merged**